snuba/query/processors/mapping_optimizer.py

```diff
@@ -159,20 +185,110 @@ def __replace_with_hash(self, condition: Expression) -> Expression:
+            they are not guaranteed to be in the same structure (but are guaranteed to be functionally equivalent)
+
+            Example:
+                ┌───┐         ┌───┐
+                │AND│         │AND│
+                ├──┬┘         └┬──┤
+                │  │           │  │
+             ┌──┴┐ c           a ┌┴──┐
+             │AND│    becomes    │AND│
+             └┬─┬┘               ├──┬┘
+              │ │                │  │
+              a b                b  c
+        """
+        if not isinstance(condition, FunctionExpr):
+            return condition
+        elif condition.function_name == BooleanFunctions.OR:
+            sub_conditions = get_first_level_or_conditions(condition)
+            pruned_conditions = [
+                self._get_condition_without_redundant_checks(c, query)
+                for c in sub_conditions
+            ]
+            return combine_or_conditions(pruned_conditions)
+        elif condition.function_name == BooleanFunctions.AND:
+            sub_conditions = get_first_level_and_conditions(condition)
+            tag_eq_match_strings = set()
+            matched_tag_exists_conditions = {}
+            for condition_id, cond in enumerate(sub_conditions):
+                tag_exist_match = _tag_exists_pattern.match(cond)
+                if tag_exist_match:
+                    matched_tag_exists_conditions[condition_id] = tag_exist_match
+                eq_match = self.__optimizable_pattern.match(cond)
+                if eq_match:
+                    tag_eq_match_strings.add(eq_match.string("key"))
+            useful_conditions = []
+            for condition_id, cond in enumerate(sub_conditions):
+                tag_exist_match = matched_tag_exists_conditions.get(condition_id, None)
+                if tag_exist_match:
+                    requested_tag = tag_exist_match.string("key")
+                    if requested_tag in tag_eq_match_strings:
+                        # how often we should skip this optimization
+                        skip_rate = state.get_config(
+                            "tags_redundant_optimizer_skip_rate", 0
+                        )
+                        if isinstance(skip_rate, float) or isinstance(skip_rate, int):
+                            if random.random() >= skip_rate:
+                                query.add_experiment(
+                                    "tags_redundant_optimizer_enabled", 1
+                                )
+                                continue
+                            else:
+                                query.add_experiment(
+                                    "tags_redundant_optimizer_enabled", 0
+                                )
+                        else:
+                            # when in doubt, apply the optimization
+                            query.add_experiment("tags_redundant_optimizer_enabled", 1)
+                            continue
```

Comment:
This means the configuration is broken. 
Let's not fix it silently, either log an error or just raise an exception.